File: db/fsharp/SalesManagement.Api/Services/ProductService.fs
module SalesManagement.Api.Services.ProductService

open System
open System.Threading.Tasks
open Npgsql
open Dapper
open SalesManagement.Domain.Models
open SalesManagement.Api.Dtos
open SalesManagement.Infrastructure.Repositories.ProductRepository

/// サービスエラーの定義
type ServiceError =
    | NotFound of string
    | BusinessError of string
    | DatabaseError of string

/// 商品を作成
let createProductAsync (connectionString: string) (request: CreateProductRequest) : Task<Result<ProductResponse, ServiceError>> =
    task {
        try
            // 商品コードの重複チェック
            let! existing = findByIdAsync connectionString request.ProductCode
            match existing with
            | Some _ ->
                return Error (BusinessError $"商品コード {request.ProductCode} は既に存在します")
            | None ->
                let now = DateTime.Now
                let product : Product =
                    { ProductCode = request.ProductCode
                      ProductFormalName = request.ProductFormalName
                      ProductAbbreviation = request.ProductAbbreviation
                      ProductNameKana = request.ProductNameKana
                      ProductType = request.ProductType
                      ModelNumber = request.ModelNumber
                      SellingPrice = request.SellingPrice
                      PurchasePrice = request.PurchasePrice
                      CostOfSales = request.CostOfSales
                      TaxType = request.TaxType
                      ProductCategoryCode = request.ProductCategoryCode
                      MiscellaneousType = request.MiscellaneousType
                      InventoryManagementFlag = request.InventoryManagementFlag
                      InventoryAllocationFlag = request.InventoryAllocationFlag
                      SupplierCode = request.SupplierCode
                      SupplierBranch = request.SupplierBranch
                      CreatedAt = now
                      CreatedBy = "SYSTEM"
                      UpdatedAt = now
                      UpdatedBy = "SYSTEM" }

                do! insertAsync connectionString product
                return Ok (ProductResponse.fromEntity product)
        with
        | ex ->
            return Error (DatabaseError ex.Message)
    }

/// すべての商品を取得
let getAllProductsAsync (connectionString: string) : Task<Result<ProductResponse list, ServiceError>> =
    task {
        try
            let! products = findAllAsync connectionString
            return Ok (products |> Seq.toList |> List.map ProductResponse.fromEntity)
        with
        | ex ->
            return Error (DatabaseError ex.Message)
    }

/// ページング対応の商品一覧取得
let getProductsAsync (connectionString: string) (page: int) (size: int) : Task<Result<PageResponse<ProductResponse>, ServiceError>> =
    task {
        try
            use connection = new NpgsqlConnection(connectionString)
            do! connection.OpenAsync()

            let offset = page * size

            // ページングクエリ
            let sql = """
                SELECT
                    商品コード AS ProductCode,
                    商品正式名 AS ProductFormalName,
                    商品略称 AS ProductAbbreviation,
                    商品名カナ AS ProductNameKana,
                    商品区分 AS ProductType,
                    型番 AS ModelNumber,
                    販売単価 AS SellingPrice,
                    仕入単価 AS PurchasePrice,
                    売上原価 AS CostOfSales,
                    税区分 AS TaxType,
                    商品分類コード AS ProductCategoryCode,
                    雑区分 AS MiscellaneousType,
                    在庫管理対象フラグ AS InventoryManagementFlag,
                    在庫引当フラグ AS InventoryAllocationFlag,
                    仕入先コード AS SupplierCode,
                    仕入先枝番 AS SupplierBranch,
                    作成日時 AS CreatedAt,
                    作成者名 AS CreatedBy,
                    更新日時 AS UpdatedAt,
                    更新者名 AS UpdatedBy
                FROM 商品マスタ
                ORDER BY 商品コード
                LIMIT @Size OFFSET @Offset
            """

            let! products = connection.QueryAsync<Product>(sql, {| Size = size; Offset = offset |})

            // 総件数取得
            let countSql = "SELECT COUNT(*) FROM 商品マスタ"
            let! total = connection.ExecuteScalarAsync<int64>(countSql)

            let content = products |> Seq.toList |> List.map ProductResponse.fromEntity
            let response = PageResponse.create content page size total

            return Ok response
        with
        | ex ->
            return Error (DatabaseError ex.Message)
    }

/// IDで商品を取得
let getProductByIdAsync (connectionString: string) (productCode: string) : Task<Result<ProductResponse, ServiceError>> =
    task {
        try
            let! product = findByIdAsync connectionString productCode
            match product with
            | Some p ->
                return Ok (ProductResponse.fromEntity p)
            | None ->
                return Error (NotFound $"商品コード {productCode} が見つかりません")
        with
        | ex ->
            return Error (DatabaseError ex.Message)
    }

/// 商品を更新
let updateProductAsync (connectionString: string) (productCode: string) (request: UpdateProductRequest) : Task<Result<ProductResponse, ServiceError>> =
    task {
        try
            let! existing = findByIdAsync connectionString productCode
            match existing with
            | None ->
                return Error (NotFound $"商品コード {productCode} が見つかりません")
            | Some product ->
                // 更新項目の適用
                let updated : Product =
                    { product with
                        ProductFormalName = request.ProductFormalName |> Option.defaultValue product.ProductFormalName
                        ProductAbbreviation = request.ProductAbbreviation |> Option.defaultValue product.ProductAbbreviation
                        ProductNameKana = request.ProductNameKana |> Option.defaultValue product.ProductNameKana
                        ProductType = request.ProductType |> Option.defaultValue product.ProductType
                        ModelNumber = request.ModelNumber |> Option.orElse product.ModelNumber
                        SellingPrice = request.SellingPrice |> Option.defaultValue product.SellingPrice
                        PurchasePrice = request.PurchasePrice |> Option.defaultValue product.PurchasePrice
                        CostOfSales = request.CostOfSales |> Option.defaultValue product.CostOfSales
                        TaxType = request.TaxType |> Option.defaultValue product.TaxType
                        ProductCategoryCode = request.ProductCategoryCode |> Option.defaultValue product.ProductCategoryCode
                        MiscellaneousType = request.MiscellaneousType |> Option.defaultValue product.MiscellaneousType
                        InventoryManagementFlag = request.InventoryManagementFlag |> Option.defaultValue product.InventoryManagementFlag
                        InventoryAllocationFlag = request.InventoryAllocationFlag |> Option.defaultValue product.InventoryAllocationFlag
                        SupplierCode = request.SupplierCode |> Option.orElse product.SupplierCode
                        SupplierBranch = request.SupplierBranch |> Option.orElse product.SupplierBranch
                        UpdatedAt = DateTime.Now
                        UpdatedBy = "SYSTEM" }

                do! updateAsync connectionString updated
                return Ok (ProductResponse.fromEntity updated)
        with
        | ex ->
            return Error (DatabaseError ex.Message)
    }

/// 商品を削除
let deleteProductAsync (connectionString: string) (productCode: string) : Task<Result<unit, ServiceError>> =
    task {
        try
            let! existing = findByIdAsync connectionString productCode
            match existing with
            | None ->
                return Error (NotFound $"商品コード {productCode} が見つかりません")
            | Some _ ->
                do! deleteAsync connectionString productCode
                return Ok ()
        with
        | ex ->
            return Error (DatabaseError ex.Message)
    }

File: db/csharp/SalesManagement.Api/Services/ProductService.cs
using System.Data;
using Dapper;
using MySql.Data.MySqlClient;
using Npgsql;
using SalesManagement.Api.Dtos;
using SalesManagement.Api.Exceptions;
using SalesManagement.Domain.Models;
using SalesManagement.Infrastructure.Repositories;

namespace SalesManagement.Api.Services;

public class ProductService
{
    private readonly ProductRepository _productRepository;
    private readonly string _connectionString;
    private readonly string _databaseType;

    public ProductService(ProductRepository productRepository, IConfiguration configuration)
    {
        _productRepository = productRepository;
        _databaseType = configuration["DatabaseType"] ?? "PostgreSQL";
        _connectionString = configuration.GetConnectionString(_databaseType)
            ?? throw new InvalidOperationException($"接続文字列が設定されていません: {_databaseType}");
    }

    private IDbConnection CreateConnection()
    {
        return _databaseType switch
        {
            "MySQL" => new MySqlConnection(_connectionString),
            "PostgreSQL" => new NpgsqlConnection(_connectionString),
            _ => throw new InvalidOperationException($"未対応のデータベース: {_databaseType}")
        };
    }

    /// <summary>
    /// 商品を作成
    /// </summary>
    public async Task<ProductResponse> CreateProductAsync(CreateProductRequest request)
    {
        // ビジネスルール: 販売単価 >= 売上原価
        if (request.UnitPrice < request.PrimeCost)
        {
            throw new BusinessException("販売単価が売上原価より低い設定はできません");
        }

        // 商品コードの重複チェック
        var existing = await _productRepository.FindByIdAsync(request.ProductCode);
        if (existing != null)
        {
            throw new BusinessException($"商品コード {request.ProductCode} は既に存在します");
        }

        var product = new Product
        {
            ProductCode = request.ProductCode,
            ProductFormalName = request.FullName,
            ProductAbbreviation = request.Name,
            ProductNameKana = request.KanaName,
            SellingPrice = request.UnitPrice,
            CostOfSales = request.PrimeCost,
            SupplierCode = request.SupplierCode,
            ProductType = "PRODUCT", // デフォルト値
            ProductCategoryCode = "CAT001", // デフォルト値
            PurchasePrice = request.PrimeCost, // 仕入単価 = 売上原価
            TaxType = 1, // デフォルト値（課税）
            MiscellaneousType = 0, // デフォルト値（通常商品）
            InventoryManagementFlag = 1, // デフォルト値（在庫管理対象）
            InventoryAllocationFlag = 0, // デフォルト値（在庫引当しない）
            SupplierBranch = 0, // デフォルト値
            CreatedAt = DateTime.Now,
            CreatedBy = "SYSTEM", // 実際は認証ユーザーから取得
            UpdatedAt = DateTime.Now,
            UpdatedBy = "SYSTEM"
        };

        await _productRepository.InsertAsync(product);
        return ProductResponse.FromEntity(product);
    }

    /// <summary>
    /// すべての商品を取得
    /// </summary>
    public async Task<List<ProductResponse>> GetAllProductsAsync()
    {
        var products = await _productRepository.FindAllAsync();
        return products.Select(ProductResponse.FromEntity).ToList();
    }

    /// <summary>
    /// ページング対応の商品一覧取得
    /// </summary>
    public async Task<PageResponse<ProductResponse>> GetProductsAsync(int page, int size)
    {
        using var connection = CreateConnection();
        connection.Open();

        int offset = page * size;

        // ページングクエリ
        var sql = @"
            SELECT
                商品コード AS ProductCode,
                商品正式名 AS ProductFormalName,
                商品略称 AS ProductAbbreviation,
                商品名カナ AS ProductNameKana,
                商品区分 AS ProductType,
                製品型番 AS ModelNumber,
                販売単価 AS SellingPrice,
                仕入単価 AS PurchasePrice,
                売上原価 AS CostOfSales,
                税区分 AS TaxType,
                商品分類コード AS ProductCategoryCode,
                雑区分 AS MiscellaneousType,
                在庫管理対象区分 AS InventoryManagementFlag,
                在庫引当区分 AS InventoryAllocationFlag,
                仕入先コード AS SupplierCode,
                仕入先枝番 AS SupplierBranch,
                作成日時 AS CreatedAt,
                作成者名 AS CreatedBy,
                更新日時 AS UpdatedAt,
                更新者名 AS UpdatedBy
            FROM 商品マスタ
            ORDER BY 商品コード
            LIMIT @Size OFFSET @Offset";

        var products = await connection.QueryAsync<Product>(sql, new { Size = size, Offset = offset });

        // 総件数取得
        var countSql = "SELECT COUNT(*) FROM 商品マスタ";
        var total = await connection.ExecuteScalarAsync<int>(countSql);

        var content = products.Select(ProductResponse.FromEntity).ToList();
        return new PageResponse<ProductResponse>(content, page, size, total);
    }

    /// <summary>
    /// IDで商品を取得
    /// </summary>
    public async Task<ProductResponse> GetProductByIdAsync(string productCode)
    {
        var product = await _productRepository.FindByIdAsync(productCode);
        if (product == null)
        {
            throw new ResourceNotFoundException($"商品コード {productCode} が見つかりません");
        }

        return ProductResponse.FromEntity(product);
    }

    /// <summary>
    /// 商品を更新
    /// </summary>
    public async Task<ProductResponse> UpdateProductAsync(string productCode, UpdateProductRequest request)
    {
        var existing = await _productRepository.FindByIdAsync(productCode);
        if (existing == null)
        {
            throw new ResourceNotFoundException($"商品コード {productCode} が見つかりません");
        }

        // ビジネスルール: 販売単価 >= 売上原価
        var newUnitPrice = request.UnitPrice ?? existing.SellingPrice;
        var newPrimeCost = request.PrimeCost ?? existing.CostOfSales;

        if (newUnitPrice < newPrimeCost)
        {
            throw new BusinessException("販売単価が売上原価より低い設定はできません");
        }

        // 更新項目の適用
        if (request.FullName != null) existing.ProductFormalName = request.FullName;
        if (request.Name != null) existing.ProductAbbreviation = request.Name;
        if (request.KanaName != null) existing.ProductNameKana = request.KanaName;
        if (request.UnitPrice.HasValue) existing.SellingPrice = request.UnitPrice.Value;
        if (request.PrimeCost.HasValue)
        {
            existing.CostOfSales = request.PrimeCost.Value;
            existing.PurchasePrice = request.PrimeCost.Value; // 仕入単価も更新
        }
        if (request.SupplierCode != null) existing.SupplierCode = request.SupplierCode;

        existing.UpdatedAt = DateTime.Now;
        existing.UpdatedBy = "SYSTEM";

        await _productRepository.UpdateAsync(existing);
        return ProductResponse.FromEntity(existing);
    }

    /// <summary>
    /// 商品を削除
    /// </summary>
    public async Task DeleteProductAsync(string productCode)
    {
        var existing = await _productRepository.FindByIdAsync(productCode);
        if (existing == null)
        {
            throw new ResourceNotFoundException($"商品コード {productCode} が見つかりません");
        }

        await _productRepository.DeleteAsync(productCode);
    }
}
